tools/repack-kernel

```diff
@@ -0,0 +1,235 @@
+    local target="$1"
+
+    if [[ -z "$target" ]]; then
+        echo "usage: repack <extract-tree>"
+        exit 1
+    fi
+
+    target=$(realpath "$target")
+    kver="$(get_kver "$target/kernel")"
+
+    (
+        # all the skeleton edits go to a local copy of distro directory
+        skeletondir="$target/skeleton"
+
+        cd "$target/work"
+        # XXX: need to be careful to build an initrd using the right kernel
+        # modules from the unpacked initrd, rather than the host which may be
+        # running a different kernel
+        (
+            # accommodate assumptions about tree layout, use the unpacked initrd
+            # to pick up the right modules
+            cd unpacked-initrd/main
+            ubuntu-core-initramfs create-initrd \
+                                  --kernelver "$kver" \
+                                  --skeleton "$skeletondir" \
+                                  --feature main \
+                                  --kerneldir "lib/modules/$kver" \
+                                  --output "$target/work/repacked-initrd"
+        )
+
+        # assumes all files are named <name>-$kver
+        ubuntu-core-initramfs create-efi \
+                              --kernelver "$kver" \
+                              --initrd repacked-initrd \
+                              --kernel vmlinuz \
+                              --output repacked-kernel.efi
+
+        cp "repacked-kernel.efi-$kver" "$target/kernel/kernel.efi"
+
+        # XXX: needed?
+        chmod +x "$target/kernel/kernel.efi"
+    )
+}
+
+cull_firmware() {
+    local target="$1"
+
+    if [[ -z "$target" ]]; then
+        echo "usage: cull-firmware <extract-tree>"
+        exit 1
+    fi
+    (
+        # XXX: drop ~450MB+ of firmware which should not be needed in under qemu
+        # or the cloud system
+        cd "$target/kernel"
+        rm -rf firmware/*
+    )
+}
+
+cull_modules() {
+    local target="$1"
+
+    if [[ -z "$target" ]]; then
+        echo "usage: cull-modules <extract-tree>"
+        exit 1
+    fi
+
+    target=$(realpath "$target")
+    kver="$(get_kver "$target/kernel")"
+
+    (
+        cd "$target/kernel"
+        # drop unnecessary modules
+        awk '{print $1}' <  /proc/modules  | sort > "$target/work/current-modules"
+        #shellcheck disable=SC2044
+        for m in $(find modules/ -name '*.ko'); do
+            noko=$(basename "$m"); noko="${noko%.ko}"
+            if echo "$noko" | grep -f "$target/work/current-modules" -q ; then
+                echo "keeping $m - $noko"
+            else
+                rm -f "$m"
+            fi
+        done
+
+        # depmod assumes that /lib/modules/$kver is under basepath
+        mkdir -p fake/lib
+        ln -s "$PWD/modules" fake/lib/modules
+        depmod -b "$PWD/fake" -A -v "$kver"
+        rm -rf fake
+    )
+}
+
+pack() {
+    local target="$1"
+
+    if [[ -z "$target" ]]; then
+        echo "usage: pack <extract-tree> <opts to snap pack>"
```

Comment:
the echo always start with the tool name like this
echo "repack-kernel: target value is required, exiting..."